### PR TITLE
[Layout Foundations] Rebuild `DropZone` with layout primitives

### LINKED ISSUE #7840: [Layout foundations] Rebuild `DropZone` with layout primitives

### PR DESCRIPTION
**Pull request by chazdean**
### WHY are these changes introduced?

Fixes #7840 

### WHAT is this pull request doing?

Refactors `DropZone` and its child components to use our layout primitives.

#### Issues with current PR ⚠️

   <details>
      <summary>Unable to center overlay text - SOLVED ✅</summary>
<table>
  <tr>
    <td>Local</td>
  </tr>
  <tr>
    <td>Screenshot 2023-01-31 at 9 42 08 AM</td>
  </tr>
  <tr>
    <td>Prod</td>
  </tr>
  <tr>
    <td>Screenshot 2023-01-31 at 9 41 54 AM</td>
  </tr>
</table>
    </details>

<details>
      <summary>Shows movement on reload (most noticeable on small and medium size dropzone)- SOLVED ✅</summary>
      

I suspect it has something to do with this style class not working properly 
```jsx
.measuring {
  visibility: hidden;
  min-height: 0;
}
```
 </details>

### How to 🎩

🖥 [Local development instructions](https://github.com/Shopify/polaris/blob/main/README.md#local-development)
🗒 [General tophatting guidelines](https://github.com/Shopify/polaris/blob/main/documentation/Tophatting.md)
📄 [Changelog guidelines](https://github.com/Shopify/polaris/blob/main/.github/CONTRIBUTING.md#changelog)

<details>
<summary>Copy-paste this code in <code>playground/Playground.tsx</code>:</summary>

```jsx
import React from 'react';
import {Page} from '../src';

export function Playground() {
  return (
    <Page title="Playground">
      {/* Add the code you want to test in here */}
    </Page>
  );
}
```

</details>

### 🎩 checklist

- [ ] Tested on [mobile](https://github.com/Shopify/polaris/blob/main/documentation/Tophatting.md#cross-browser-testing)
- [ ] Tested on [multiple browsers](https://help.shopify.com/en/manual/shopify-admin/supported-browsers)
- [ ] Tested for [accessibility](https://github.com/Shopify/polaris/blob/main/documentation/Accessibility%20testing.md)
- [ ] Updated the component's `README.md` with documentation changes
- [ ] [Tophatted documentation](https://github.com/Shopify/polaris/blob/main/documentation/Tophatting%20documentation.md) changes in the style guide

### CODE REVIEW

File: polaris-react/src/components/Box/Box.tsx [hunk 1]
```diff
@@ -117,6 +117,10 @@ export interface BoxProps extends React.AriaAttributes {
   minWidth?: string;
   /** Maximum width of container */
   maxWidth?: string;
+  /** onDrag */
+  onDragStart?: React.DragEventHandler<HTMLElement>;
+  /** onClick */
+  onClick?: React.MouseEventHandler<HTMLElement>;
```
Comment (yurm04) [inline polaris-react/src/components/Box/Box.tsx]: i've been out for a bit so take this comment with a grain of salt: adding the drag (and maybe even onClick?) is a departure from our `Box` as a presentational component principle. Are there other approaches we can take that don't add these props, or are we cool with expanding the purpose of `Box` to include more use cases?

Comment (chazdean) [reply polaris-react/src/components/Box/Box.tsx]: I don't think this has come up yet, looking forward to hearing some of the feedback around this as I was also unsure of the best approach in regards to handling the drag and onClick functionality

Comment (aveline) [reply polaris-react/src/components/Box/Box.tsx]: Let's avoid adding the `drag` and `onClick` to `Box` at this time. We should keep our focus for this project on layout and presentational components. Any elements that need interactive support are not presentational and so fall outside the scope of layout.

In a way this forces us to be aware of separation of concerns between layout and interactive elements.

File: polaris-react/src/components/DropZone/DropZone.tsx [hunk 2]
```diff
@@ -390,24 +391,28 @@ export const DropZone: React.FunctionComponent<DropZoneProps> & {
         action={labelAction}
         labelHidden={labelHiddenValue}
       >
-        <div
+        <Box
```
Comment (aveline) [inline polaris-react/src/components/DropZone/DropZone.tsx]: What you could try here is separating the visual styles from the interactive functionality by having both the `Box` for the visual and layout styles and having the `div` for the interactive functionality.

**Review by aveline — CHANGES_REQUESTED** [review]
Nice to see `AlphaStack` replace all those instances of `Stack` 🎉

File: polaris-react/src/components/Box/Box.tsx (hunk 1 above)
Comment (chazdean) [reply polaris-react/src/components/Box/Box.tsx]: That makes sense 👍🏾

**Review by nayeobkim — APPROVED** [review]
Reviewed Chromatic and Storybook—all visually looked good to me

File: polaris-react/src/components/DropZone/DropZone.tsx [hunk 3]
```diff
@@ -419,14 +449,32 @@ export const DropZone: React.FunctionComponent<DropZoneProps> & {
...
+            {(size === 'medium' || size === 'large') && (
+              <Text
+                variant="bodySm"
+                as="p"
+                fontWeight="bold"
+                {...((internalError || error) && {
+                  color: 'critical',
+                })}
+              >
+                <div className={styles.overlayContainer}>{text}</div>
```
Comment (aveline) [inline polaris-react/src/components/DropZone/DropZone.tsx]: Can `.overlayContainer` be removed?

Comment (chazdean) [reply polaris-react/src/components/DropZone/DropZone.tsx]: Yes! Forget to take that out

**Review by laurkim — APPROVED** [review]

**Review by aveline — APPROVED** [review]

**Pull request merged**